service/notification/internal/app/scheduler/scheduler.go

```diff
@@ -0,0 +1,169 @@
+package scheduler
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-boredgus/service/notification/internal/app"
+	messages "github.com/GenesisEducationKyiv/software-engineering-school-4-0-boredgus/service/notification/internal/broker/gen"
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-boredgus/service/notification/internal/config"
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-boredgus/service/notification/internal/entities"
+	"github.com/robfig/cron/v3"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type JobSpec struct {
+	Mins  int
+	Hours int
+}
+
+func NewJobSpec(t time.Time) JobSpec {
+	return JobSpec{Mins: t.Minute(), Hours: t.Hour()}
+}
+func (s JobSpec) String() string {
+	return fmt.Sprintf("%v %v * * *", s.Mins, s.Hours)
+}
+
+type (
+	ScheduledDispatch struct {
+		Data    *entities.Dispatch
+		EntryID cron.EntryID
+		Spec    JobSpec
+	}
+
+	Publisher interface {
+		PublishAsync(subject string, payload []byte) error
+	}
+	Converter interface {
+		Convert(ctx context.Context, baseCcy string, targetCcies []string) (map[string]float64, error)
+	}
+
+	dispatchScheduler struct {
+		cron                *cron.Cron
+		mu                  *sync.Mutex
+		scheduledDispatches map[string]*ScheduledDispatch
+		broker              Publisher
+		converter           Converter
+		logger              config.Logger
+	}
+)
+
+func NewDispatchScheduler(
+	broker Publisher,
+	converter Converter,
+	logger config.Logger,
+) *dispatchScheduler {
+	return &dispatchScheduler{
+		mu:                  &sync.Mutex{},
+		cron:                cron.New(cron.WithLocation(time.UTC)),
+		scheduledDispatches: make(map[string]*ScheduledDispatch),
+	}
+}
+
+func (s *dispatchScheduler) invokeSendingOfDispatch(d *entities.Dispatch) {
+	msg := messages.SendDispatchCommand{
+		EventType: messages.EventType_SEND_DISPATCH,
+		Timestamp: timestamppb.New(time.Now().UTC()),
+		Payload: &messages.Dispatch{
+			Emails:  d.Emails,
+			BaseCcy: d.BaseCcy,
+		},
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), app.TimeoutOfProcessing)
+	defer cancel()
+
+	rates, err := s.converter.Convert(ctx, d.BaseCcy, d.TargetCcies)
+	if err != nil {
+		s.logger.Errorf("failed to get rates: %v", err)
+
+		return
+	}
+	msg.Payload.Rates = rates
+
+	marshalled, err := proto.Marshal(&msg)
+	if err != nil {
+		s.logger.Errorf("failed to marshal SendDispatchCommand: %v", err)
+
+		return
+	}
+
+	if err := s.broker.PublishAsync(app.SendDispatchCommand, marshalled); err != nil {
+		s.logger.Errorf("failed to emit SendDispatch commands: %v", err)
+	}
+}
```

Comment:
Seems like detailed job func that scheduler doesn't need to know. Also violation of SRP principle

Reply:
But scheduler just invoks this complex func instead of containing it inside

Reply:
In other way I nedd to pass invoke function as argument when subscription is created or cancelled. Does this approach satisfies us more?